### PR TITLE
JAMES-2586 Implement BlobReferenceSource(s) for postgres-app

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

**Comment by chibenwa** [comment]
Please write integration tests for bob deletion using the deduplicating blob store.

**Review by Arsnael — COMMENTED** [review]
Read it nothing to add

**Comment by hungphan227** [comment]
The only integration test I found for blob GC is RabbitMQWebAdminServerBlobGCIntegrationTest in distributed-webadmin-integration-test. Should we have integration test for this ticket covered in the ticket Webadmin integration tests for postgres-app (https://github.com/linagora/james-project/issues/5005) ?

**Comment by Arsnael** [comment]
I would be ok with it

**Review by Arsnael — APPROVED** [review]

**Comment by chibenwa** [comment]
Then a demo video so that I am sure such an important topic received the attention it requires:

```
GIVEN I start with generation 1
AND I create 2 messages with attachent A
AND I delete one message
WHEN I reboot the server on generation 2 and run GC
THEN I expect no deletion

GIVEN I start with generation 1
AND I create 2 messages with attachent A
AND I delete two message
WHEN I reboot the server on generation 2 and run GC
THEN I expect deletion of attachent A content
```

BTW it is not too late to write such an integration test ;-)

**Comment by chibenwa** [comment]
> RabbitMQWebAdminServerBlobGCIntegrationTest in distributed-webadmin-integration-test

Would do of course.

But I need to see a github ticket about it so that is is not forgotten.

(and the mantra : what is not tested do not work - might be relevant here)

Time to fasten that webadmin integration test for PG app?

**Comment by hungphan227** [comment]
integration test ticket created https://github.com/linagora/james-project/issues/5031

**Comment by quantranhong1999** [comment]
> integration test ticket created https://github.com/linagora/james-project/issues/5031

https://github.com/linagora/james-project/issues/5005
Why another ticket?

**Comment by hungphan227** [comment]
> > integration test ticket created [linagora#5031](https://github.com/linagora/james-project/issues/5031)
> 
> [linagora#5005](https://github.com/linagora/james-project/issues/5005) Why another ticket?

https://github.com/orgs/linagora/projects/2/views/2?pane=issue&itemId=48767936 just implements webadmin-integration-test-common. Other tests wuould be added in another ticket.

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import org.apache.james.blob.api.BlobReferenceSource;
+import org.apache.james.mailbox.postgres.mail.dao.PostgresMessageDAO;
+
+import reactor.core.publisher.Flux;
+
+public class PostgresMessageBlobReferenceSource implements BlobReferenceSource {
+    private PostgresMessageDAO postgresMessageDAO;
+
+    @Inject
+    public PostgresMessageBlobReferenceSource(PostgresMessageDAO postgresMessageDAO) {
```
Comment (vttranlina) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:34]: Please don't inject `PostgresMessageDAO` as a singleton bean

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        this.postgresMessageDAO = postgresMessageDAO;
+    }
+
+    @Override
+    public Flux<BlobId> listReferencedBlobs() {
+        return postgresMessageDAO.listBlobs();
```
Comment (vttranlina) [inline mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: For byPassRLS
we can:
- list all domain
- for each domain, list all blobId
- concat blobId

WDYT?
Alternative idea:
- 1. Create a mid-table for storage blobId without domain columns (for bypass RLS)
- 2. Create a "super" Connection (by super User?) for bypass RLS

Comment (chibenwa) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: No Hung approach seems more straightforward to me: obtain a root session for the dedup purpose...

Comment (vttranlina) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: > have a complexity O(n*n) as for each domain

it shouldn't even if it is a task?

To create a "root session", the basic way is configuration properties support declaring an RLS user, super User, then we can create a root session based on the superuser. 
Not sure can have another easier way?

Comment (Arsnael) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: What's wrong with having a super user creating a root session? as long as it's just used internally I think we good?

Comment (vttranlina) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: > What's wrong with having a super user creating a root session? as long as it's just used internally I think we good?

No Wrong, it's just more code, I'm ok with that, it is also my think
> Alternative idea:
> Create a "super" Connection (by super User?) for bypass RLS

Comment (Arsnael) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: Ah yes fair :) The first idea is more requests to the db though to list all blobs. I would not be surprised as well that we find more cases later where a root session would be useful and save us from more sql requests too, so I don't think it's bad to invest a bit of time on it now :)

Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: I have not done anything yet in this ticket to solve this issue. I have just discussed with Tung about this to find solution. Not sure how you @chibenwa obtain root session. For me, the only way to get it is using root user or an user that have privilege of bypassing rls to obtain a session that can bypass RLS. And to achieve that, we have to create two postgresqlConnectionFactory(s):

```
new PostgresqlConnectionFactory(PostgresqlConnectionConfiguration.builder()
            .host(postgresConfiguration.getUri().getHost())
            .port(postgresConfiguration.getUri().getPort())
            .username(postgresConfiguration.getCredential().getUsername())
            .password(postgresConfiguration.getCredential().getPassword())
            .database(postgresConfiguration.getDatabaseName())
            .schema(postgresConfiguration.getDatabaseSchema())
            .build()); 
```
one PostgresqlConnectionFactory has been used so far. A new one would use root user or an user that have privilege of bypassing rls.

WDYT?

**Comment by vttranlina** [comment]
Note: I see the todo comment in the code, can you resolve it in this pr?
>   // should remove this after https://github.com/linagora/james-project/issues/4998

https://github.com/apache/james-project/blob/3e86c78b58f53e50606aea06adeb17ab2a778861/server/apps/postgres-app/src/main/java/org/apache/james/PostgresJamesServerMain.java#L142

File: mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java (hunk 2 above)
Comment (hungphan227) [reply mailbox/postgres/src/main/java/org/apache/james/mailbox/postgres/mail/PostgresMessageBlobReferenceSource.java:40]: fixed

File: backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java [hunk 3]
```diff
@@ -114,23 +114,22 @@ private void querySettingExtension() throws IOException, InterruptedException {
...
             .build();
 
         connectionFactory = new PostgresqlConnectionFactory(PostgresqlConnectionConfiguration.builder()
-            .host(postgresConfiguration.getUri().getHost())
-            .port(postgresConfiguration.getUri().getPort())
+            .host(postgresConfiguration.getHost())
+            .port(postgresConfiguration.getPort())
             .username(postgresConfiguration.getCredential().getUsername())
             .password(postgresConfiguration.getCredential().getPassword())
             .database(postgresConfiguration.getDatabaseName())
```
Comment (vttranlina) [inline backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java:135]: Should we add a new method for PostgresExtension:
`getNonRLSPostgresExecutor`?

tldr:

I see the `PostgresMessageBlobReferenceSourceTest` is tested with `withoutRowLevelSecurity` 
I guess it will not pass if enable `withRowLevelSecurity` + insert into the messages table with some random domain. Then `blobReferenceSource` will not retrieve full lists

File: server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    private static final String HEADERS_SEPARATOR = ";  ";
+
+    private final PostgresExecutor postgresExecutor;
+    private final Store<MimeMessage, MimeMessagePartsId> mimeMessageStore;
+    private final BlobId.Factory blobIdFactory;
+
+    @Inject
+    public PostgresMailRepositoryContentDAO(PostgresExecutor postgresExecutor,
+                                  MimeMessageStore.Factory mimeMessageStoreFactory,
+                                  BlobId.Factory blobIdFactory) {
```
Comment (vttranlina) [inline server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java:98]: indent

**Review by Arsnael — APPROVED** [review]

File: server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+    }
+
+    private Mail toMail(Record record, MimeMessage mimeMessage) throws MessagingException {
+        List<MailAddress> recipients = Arrays.stream(record.get(RECIPIENTS))
+            .map(Throwing.function(MailAddress::new))
+            .collect(ImmutableList.toImmutableList());
+
+        PerRecipientHeaders perRecipientHeaders = getPerRecipientHeaders(record);
+
+        List<Attribute> attributes = Hstore.hstore(record.get(ATTRIBUTES, LinkedHashMap.class))
```
Comment (vttranlina) [inline server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java:241]: I don't know why we need call to `Hstore` here
It should not invoked when un-necessary (that created for of Jooq extension)
We can entrySet directly on LinkedHashMap

```java
((LinkedHashMap<String, String>) record.get(PER_RECIPIENT_SPECIFIC_HEADERS, LinkedHashMap.class))
            .entrySet()
```

File: backends-common/postgres/src/main/java/org/apache/james/backends/postgres/PostgresConfiguration.java [hunk 6]
```diff
@@ -99,36 +162,22 @@ public Builder rowLevelSecurityEnabled() {
...
+                Preconditions.checkArgument(nonRLSPassword.isPresent() && !nonRLSPassword.get().isBlank(), "You need to specify nonRLSPassword");
+            }
 
-            return new PostgresConfiguration(postgresURI,
-                parseCredential(postgresURI),
+            return new PostgresConfiguration(host.orElse(HOST_DEFAULT_VALUE),
+                port.orElse(PORT_DEFAULT_VALUE),
                 databaseName.orElse(DATABASE_NAME_DEFAULT_VALUE),
                 databaseSchema.orElse(DATABASE_SCHEMA_DEFAULT_VALUE),
+                new Credential(username.get(), password.get()),
```
[the remark this answers is not in the document]
Comment (hungphan227) [reply backends-common/postgres/src/main/java/org/apache/james/backends/postgres/PostgresConfiguration.java:177]: I think it is not really necessary

File: backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java (hunk 3 above)
Comment (hungphan227) [reply backends-common/postgres/src/test/java/org/apache/james/backends/postgres/PostgresExtension.java:135]: We should do this in integration tests

File: server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java [hunk 7]
```diff
@@ -0,0 +1,355 @@
...
+            mail.setMessage(mimeMessage);
+        }
+
+        return mail;
+    }
+
+    private PerRecipientHeaders getPerRecipientHeaders(Record record) {
+        PerRecipientHeaders perRecipientHeaders = new PerRecipientHeaders();
+
+        Hstore.hstore(record.get(PER_RECIPIENT_SPECIFIC_HEADERS, LinkedHashMap.class))
```
Comment (vttranlina) [inline server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java:273]: Same https://github.com/apache/james-project/pull/1900#discussion_r1445716347

File: server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java (hunk 5 above)
Comment (quantranhong1999) [reply server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java]: The `ATTRIBUTES` data type is `hstore`. 
`record.get(ATTRIBUTES)` return `Hstore` at compilation time, but return `LinkedHashMap` at run time. Funny, no? (jooq bug I bet)

That is why I casted `LinkedHashMap.class` there.

I see nothing else wrong with that.

Comment (quantranhong1999) [reply server/data/data-postgres/src/main/java/org/apache/james/mailrepository/postgres/PostgresMailRepositoryContentDAO.java]: But I am OK to remove the intermediate Hstore object there.

**Comment by vttranlina** [comment]
CI fail
```
shouldPreserveDsnParameters – org.apache.james.mailrepository.postgres.PostgresMailRepositoryTest

javax.mail.internet.AddressException: Out of data at position 1 in 'dsn-ret' in string ``dsn-ret'' at position 1
Stacktrace
com.github.fge.lambdas.ThrownByLambdaException: javax.mail.internet.AddressException: Out of data at position 1 in 'dsn-ret' in string ``dsn-ret'' at position 1
	at com.github.fge.lambdas.supplier.ThrowingSupplier.get(ThrowingSupplier.java:27)
	at org.apache.james.mailrepository.postgres.PostgresMailRepositoryContentDAO.asMailAddress(PostgresMailRepositoryContentDAO.java:297)
	at org.apache.james.mailrepository.postgres.PostgresMailRepositoryContentDAO.lambda$recipientToHeaderStream$16(PostgresMailRepositoryContentDAO.java:291)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
```

__________
Don't forget update the document for new postgres properties

**Comment by hungphan227** [comment]
> CI fail
> 
> ```
> shouldPreserveDsnParameters – org.apache.james.mailrepository.postgres.PostgresMailRepositoryTest
> 
> javax.mail.internet.AddressException: Out of data at position 1 in 'dsn-ret' in string ``dsn-ret'' at position 1
> Stacktrace
> com.github.fge.lambdas.ThrownByLambdaException: javax.mail.internet.AddressException: Out of data at position 1 in 'dsn-ret' in string ``dsn-ret'' at position 1
> 	at com.github.fge.lambdas.supplier.ThrowingSupplier.get(ThrowingSupplier.java:27)
> 	at org.apache.james.mailrepository.postgres.PostgresMailRepositoryContentDAO.asMailAddress(PostgresMailRepositoryContentDAO.java:297)
> 	at org.apache.james.mailrepository.postgres.PostgresMailRepositoryContentDAO.lambda$recipientToHeaderStream$16(PostgresMailRepositoryContentDAO.java:291)
> 	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
> 	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1655)
> 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
> 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
> 	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
> ```
> 
> Don't forget update the document for new postgres properties

where is the document?

**Comment by hungphan227** [comment]
> > where is the document?
> 
> sorry, I rechecked it again, now we haven't documented for it, just add a new property and comment in the sample configuration file https://github.com/apache/james-project/blob/c580f1db2284f380b1171633820512cd10e70fb3/server/apps/postgres-app/sample-configuration/postgres.properties

added

**Review by vttranlina — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Comment by chibenwa** [comment]
Please rebase
